vignettes/query.Rmd

```diff
@@ -0,0 +1,212 @@
+
+There are four other functions
+
+* `latest(expr)` finds the latest packet satisfying `expr` - it always returns a length 1 character, but this is `NA_character_` if no suitable packet is found. If no `expr` is given then the latest of all packets is returned.
+* `single(expr)` is like `latest(expr)` except that it is an error if `expr` does not evaluate to exactly one packet id
+* `usedby(expr, FALSE)` where `expr` is either a literal `id` or an expression which returns 1 `id`. This finds all packets which were used in generation of packet with `id` returned from `expr` (see dependencies section for more details).
+
+## Special simple queries
+
+There are two shorthand queries:
+
+* `latest` is equivalent to `latest()` (most useful when applied with a scope)
+* a string matching the regular expression for an id (`^([0-9]{8}-[0-9]{6}-[[:xdigit:]]{8})$`) is equivalent to `single(id == "<id>")` where `"<id>"` is the string provided)
+
+## Scoping queries
+
+**WARNING: we may remove this**
+
+Scoping queries can be used as a shorthand for filtering the returned packets. In the future they could be used to reduce the set of packets that are searched over to speed up query evaluation. They join together with the main query as `(scope) && (expr)`, except when the `expr` is a call to `latest` or `single`. In this case they combine as `latest((scope) && (expr))` or `single((scope) && (expr))`. This is useful if you want to limit the search to a particular name or location but perform some more detailed search.
+
+For example, the query
+
+```r
+outpack_query(quote(parameter:x == 1), scope = quote(name == "data"))
+```
+
+is equivalent to
+
+```r
+outpack_query(quote(parameter:x == 1 && name == "data"))
+```
+
+`orderly2` uses this functionality when resolving dependencies with `orderly2::orderly_dependency`.
+
+### Scoping on name
+
+Very often users will want to scope by name so instead of passing `scope` argument there is a shorthand `name` argument for usability.
+
+```r
+outpack_query(quote(parameter:x == 1), name = "data")
+```
+
+Which is the equivalent of
+
+```r
+outpack_query(quote(parameter:x == 1), scope = quote(name == "data"))
+```
+
+## Dependencies
+
+If we have 2 packets, where B depends on output from A (i.e. we call `outpack_packet_use_dependency(id_a, ...)` when running packet B) we can draw this as.
+
+```{r, echo = FALSE, fig.align = 'center'}
+knitr::include_graphics("images/dependency_terminology.png")
+```
+
+We could equivalently say
+
+* A is a dependency of B, B depends on A
+* A is used by B , B uses A
+* A is a parent of B, B is a child of A
+
+With the tree of dependencies built by outpack we might want to search for packets which have been used by another packet. We can use the query function `usedby(id)` to list all packets which are used by `id`. This will search recursively through all packets used by `id` and its parents and its parents parents and so on.
```

Comment:
```suggestion
With the tree of dependencies built by outpack we might want to search for packets which have been used by another packet. We can use the query function `usedby(id)` to list all packets which are used by `id`. This will search recursively through all packets used by `id` and its parents and its parents' parents and so on.
```